quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ExportPropertyMappers.java

```diff
@@ -49,43 +53,70 @@ public static PropertyMapper<?>[] getMappers() {
                         .build(),
                 fromOption(ExportOptions.USERS)
                         .to("kc.spi-export-dir-users-export-strategy")
+                        .isEnabled(ExportPropertyMappers::isDirProvider)
                         .paramLabel("strategy")
                         .build(),
                 fromOption(ExportOptions.USERS_PER_FILE)
                         .to("kc.spi-export-dir-users-per-file")
+                        .isEnabled(ExportPropertyMappers::isDirProvider)
                         .paramLabel("number")
                         .build()
         };
     }
 
+    private static final Option<String> EXPORTER_PLACEHOLDER = new OptionBuilder<>("exporter", String.class)
+            .category(OptionCategory.EXPORT)
+            .description("Placeholder for determining export mode")
+            .buildTime(false)
+            .hidden()
+            .build();
+
+    public static boolean isProperExporter() {
+        return Configuration.getOptionalValue(EXPORTER_PROPERTY).isPresent();
+    }
+
+    private static boolean isSingleFileProvider() {
+        return isProvider(SINGLE_FILE);
+    }
+
+    private static boolean isDirProvider() {
+        return isProvider(DIR);
+    }
+
+    private static boolean isProvider(String provider) {
+        return Configuration.getOptionalValue(EXPORTER_PROPERTY)
+                .filter(provider::equals)
+                .isPresent();
+    }
+
+    private static final String SINGLE_FILE = "singleFile";
+    private static final String DIR = "dir";
+
     private static Optional<String> transformExporter(Optional<String> option, ConfigSourceInterceptorContext context) {
-        ConfigValue exporter = context.proceed("kc.spi-export-exporter");
+        ConfigValue exporter = context.proceed(EXPORTER_PROPERTY);
         if (exporter != null) {
             return Optional.of(exporter.getValue());
         }
-        if (option.isPresent()) {
-            return Optional.of("singleFile");
-        }
-        ConfigValue dirConfigValue = context.proceed("kc.spi-export-dir-dir");
-        if (dirConfigValue != null && dirConfigValue.getValue() != null) {
-            return Optional.of("dir");
-        }
-        ConfigValue dirValue = context.proceed("kc.dir");
-        if (dirValue != null && dirValue.getValue() != null) {
-            return Optional.of("dir");
-        }
-        if (System.getProperty(PROVIDER) == null) {
-            throw new CommandLine.PicocliException("Must specify either --dir or --file options.");
-        }
-        return Optional.empty();
+
+        var file = Configuration.getOptionalValue("kc.spi-export-single-file-file").map(f -> SINGLE_FILE);
+        var dir = Configuration.getOptionalValue("kc.spi-export-dir-dir")
+                .or(() -> Configuration.getOptionalValue("kc.dir"))
+                .map(f -> DIR);
+
+        // Only one option can be specified
+        boolean xor = file.isPresent() ^ dir.isPresent();
```

Comment:
PropertyMappers now also support validators. In this case the actual validation seems to be performed up in the command classes, but may be possible to move it onto the Import/ExportPropertyMappers logic. Not sure that it would really matter - it's just a question of where the logic appears.

Reply:
It's a little bit more complicated. It's not possible to use the present validation logic in PropertyMapper. The thing is, that the ConfigValue in the `validateConfig` is obtained from the `.getFrom()` property, where the transformer is not executed, thus, the value is not computed, thus null. 

The validation is skipped for mappers, where the value is null. 

In order to execute additional validation, even when the value is null, we should do sth like this:
https://github.com/mabartos/keycloak/commit/2b882a302dc61b749c963b0ab8bb9f7ef49f781d

Probably, there might be a better solution, but I'd also rather solve this case in a follow-up task as it's little bit out of scope of this PR. At least, I put the validation logic to the mappers class and referenced from the command validation method. 

@shawkins Does it work for you, to resolve it in a follow-up task?

Reply:
Created follow-up task: https://github.com/keycloak/keycloak/issues/27025